### PR TITLE
Add dedicated [Arb/Acb]Ref[Vector/Matrix] type

### PR DESCRIPTION
**Pull request by saschatimme**
This tries to resolve the discussion in #44 .

* `[Arb/Acb][Vector/Matrix]` now returns an honest `Arb` (resp. `Acb`)
* `[Arb/Acb]Ref[Vector/Matrix]` is a new type which returns an `ArbRef` (resp. `AcbRef`)
* Possible to convert between these two container formats without allocations 
* Introduces new function `ref(A, i [, j])` to index `[Arb/Acb][Vector/Matrix]` and obtain `ArbRef` (resp. `AcbRef`).

I hope this addresses all concerns @kalmarek  @Joel-Dahne .

### CODE REVIEW

**Review by kalmarek — APPROVED** [review]
It's probably too much code for a quick review I did, but let's not wait with this
@Joel-Dahne If you don't have major complaints, let's merge this now and iterate on it later

**Comment by Joel-Dahne** [comment]
Go ahead and merge!

**Comment by kalmarek** [comment]
ok, I have one complaint:) `ArbMatrix(::AbstractMatrix, prec=...)` uses default argument and not a `prec` kwarg as it is everywhere else...

**Comment by saschatimme** [comment]
Oh yes this I also noticed already,  but I didn’t want to sneak this into this PR.  But I can obviously do this :)

File: src/vector.jl [hunk 1]
```diff
@@ -62,20 +105,19 @@ end
 
 # Arithmetic
 for (jf, af) in [(:+, :add!), (:-, :sub!)]
-    @eval function Base.$jf(v::T, w::T) where {T<:Union{ArbVector,AcbVector}}
+    @eval function Base.$jf(v::T, w::T) where {T<:Vectors}
```
Comment (kalmarek) [inline src/vector.jl:108]: these functions does not even work on the syntactic level ;)
but it's too much work to fix it here, I already did in #56

File: src/matrix.jl [hunk 2]
```diff
@@ -90,24 +125,25 @@ for T in [:ArbMatrix, :AcbMatrix]
...
     set!(A, B)
     A
 end
 
-## Common methods
-
 # Basic arithmetic
 for (jf, af) in [(:+, :add!), (:-, :sub!)]
-    @eval function Base.$jf(A::T, B::T) where {T<:Union{ArbMatrix,AcbMatrix}}
+    @eval function Base.$jf(A::T, B::T) where {T<:Matrices}
```
Comment (kalmarek) [inline src/matrix.jl:146]: please add one test for each of the functions

File: src/types.jl [hunk 3]
```diff
@@ -115,55 +115,107 @@ function Acb(x::AcbRef; prec::Integer = precision(x))
...
+AcbRefMatrix(r::Integer, c::Integer; prec::Integer = DEFAULT_PRECISION[]) =
+    AcbRefMatrix(acb_mat_struct(r, c), prec)
+
+# conversion between ref and non-ref arrays.
+for T in [:Arb, :Acb], A in [:Vector, :Matrix]
+    TA = Symbol(T, A)
+    TRefA = Symbol(T, :Ref, A)
+    @eval begin
+        $TRefA(M::$TA) = $TRefA(cstruct(M), precision(M))
+        $TA(M::$TRefA) = $TA(cstruct(M), precision(M))
```
Comment (kalmarek) [inline src/types.jl:180]: please add one tests for each constructors

**Review by kalmarek — CHANGES_REQUESTED** [review]
we need to make sure the functions run as defined ;)

**Review by kalmarek — APPROVED** [review]

**Comment by kalmarek** [comment]
yay!

**Pull request merged**